src/lem/circuit.rs

```diff
@@ -0,0 +1,890 @@
+//! # Constraint system for LEM
+//!
+//! Here we describe how we generate bellperson constraints for LEM, such that
+//! it can be used with Nova folding to implement the Lurk evaluation.
+//!
+//! ## LEM paths and the implications:
+//!
+//! LEM implements logical branching using some specific LEMOPs such as `MatchTag`
+//! and `MatchSymbol`. By nesting them we create a set of paths that interpretation
+//! can follow. We call them **virtual** and **concrete** paths.
+//! In particular, the followed path is the concrete one. We use a Boolean
+//! variable to indicate whether a path is concrete or not. This allows us to
+//! construct an **implication system**, which is responsible for ensuring that
+//! allocated variables in the concrete path are equal to their expected values
+//! but such equalities on virtual paths are irrelevant.
+//!
+//! ## Slot optimization:
+//!
+//! Some operations like Poseidon hash can be relatively expensive in the circuit,
+//! therefore we want to avoid wasting constraints with them as much as possible.
+//! In order to achieve this goal, we only add constraints to a limited number of
+//! *slots* that correspond to a maximum number of such expensive operations that
+//! may occurr in a single interpretation round. This optimization avoids
+//! constraining every expensive operation on every possible path.
+//!
+//! Shortly, we construct the hash slot system using the next steps:
+//!
+//! * STEP 1: Static analysis, when we traverse LEM for the first time and allocate
+//! slots for hashes in all virtual paths. We only add hashes that where not yet
+//! previously added, therefore ensuring deduplication of hashes in different
+//! branches.
+//!
+//! * STEP 2: During interpretation (second traversal) we gather information
+//! related to each visited slot.
+//!
+//! * STEP 3: During construction of constraints, we first preallocate the preimage
+//! and image for each slot; then we synthesize all slots; and finally we traverse
```

Comment:
image -> hash

Reply:
What if the slot type is `SlotType::BitDecomp` (in the future)?

Reply:
I see… if this is intended to be fully general then maybe it's okay. However, in that case, you should not refer to 'hashes' in Step 1 either. That said, I think you're only dealing with hashes now… I would probably rather see this work avoid premature abstraction in its language — given how hard it has been so far to converge (witness the inconsistency in the abstraction of the step descriptions just noted).

Reply:
I'm trying to avoid the need to come back and change the occurrences of "hash" when we meant "image" more generally. I have an idea: use "hash" when it's about hashing and "image" when we want to be more general. Is that okay?

Reply:
Maybe? It's confused because you're mixing levels of abstraction already. You're referring to 'hashing' (which is 'fine' because you don't do anything else yet) — but then using the 'generic' terminology. It's made more confusing since you've chosen generic terminology that contains a term ('preimage') that is also used in one of its more specific instantiations.

But in the general case, it's far from clear that image/preimage are excellent choices, especially because the rhyme with hashing is unavoidable (because of 'preimage').

For example, in the case of bit decomposition, will you have the `Num` be the preimage and the allocated bits be the image? That would be consistent with calling the operation 'bit decomposition' as a structural equivalent of 'hashing'. But it won't have the same 'shape' as a function — since it will be one->many instead of many->one.

I would just choose generic language that doesn't try to cleverly rhyme with the hashing language, since that punning is in fact a source of confusion.

Reply:
What about `inputs`/`output`? The plural/singular already helps with some disambiguation regarding the one->many vs many->one conflict. Plus, those words have the same number of characters